### ISSUE TITLE
Event handler lookup inconsistency of XamlC, XamlLoader and WPF

### ISSUE THREAD

**Issue by akihikodaki** [xamarin/Xamarin.Forms#5100]
### Description
For example, consider the following case:
* `C` is a class.
* `D` is a class inheriting `C`.
* There is a XAML with `C` as `/@x:Class`.

If you compile the XAML and calls `D.InitializeComponent`, it searches `C` for an event handler ignoring inherited methods. However, if you load the XAML on `D` with `Xamarin.Forms.Xaml.Internals.XamlLoader`, it searches `D` for an event handler, ignoring inherited methods, too.

If I read the source code correctly, WPF's XAML loader seems to search the given value's type and its superclasses for an event handler. (So it searches both of `C` and `D` in the example.) However, I'm not sure if it is desirable because it may find a private method in a superclass.

I hope it behaves like `System.Reflection.BindingFlags.FlattenHierarchy` rooting the class specified by `x:Class`, but I think it depends on your design decision.

### Steps to Reproduce

Compare the behavior of XamlC and XamlLoader in the case presented in the description.

### Expected Behavior

XamlC and XamlLoader behaves in the same manner.

### Actual Behavior

See the description.

### Basic Information

- Version with issue: `3.4.0-sr2`
- Last known good version: N/A
- IDE: Visual Studio Code
- Platform Target Frameworks: 
  - iOS:  N/A 
  - Android: N/A  
  - UWP:  N/A  
- Android Support Library Version: N/A 
- Nuget Packages: N/A
- Affected Devices: Mono on GNU/Linux

### Screenshots

N/A

### Reproduction Link

https://github.com/akihikodaki/Xamarin.Forms/tree/gh5100

**Comment by kingces95** [comment]
Could you attach a small reproduction, please?

**Comment by akihikodaki** [comment]
I found the description of this issue is invalid, particularly the following statement:
> However, if you load the XAML on `D` with `Xamarin.Forms.Xaml.Internals.XamlLoader`, it searches `D` for an event handler, ignoring inherited methods, too.

I cannot reproduce this on Mono 5.16.0. I suspect the cause is the runtime which I used when I encountered this problem. (Unity 2018.3.0f2, which apparently has Mono 5.11.0: https://github.com/Unity-Technologies/mono/tree/unity-2018.3-mbe)

I think this issue is invalid so I'll close this issue.

**Comment by akihikodaki** [comment]
I'll reopen this issue because I managed to reproduce in a slightly scenario altered from the real case. I added a link to the reproduction.

**Comment by kingces95** [comment]
Can you include steps to reproduce and the actual and expected behaviors?

**Comment by akihikodaki** [comment]
> Is that what you intended?

Yes, it has a test case reproducing the inconsistency.

**Comment by samhouts** [comment]
@StephaneDelcroix Would you like to comment on this one? Thanks!

**Comment by StephaneDelcroix** [comment]
@kingces95 
> Think link is to a Xamarin branch. Is that what you intended?

this is a link to a branch, on which the HEAD commit is the actual test https://github.com/akihikodaki/Xamarin.Forms/commit/2c00b421ddf0412f063cf81e5f3416647f274ca9

@samhouts 
> @StephaneDelcroix Would you like to comment on this one? Thanks!

Of course: @akihikodaki is usually right. we at least want consistency between our 2 loaders, and ideally with WPF. Reflection calls can be slightly different between netstandard1.0 and 2.0 variations, and that what might cause it

Looking at the test now. @akihikodaki you expect the test to fail because the signature is wrong ? let me try to repro this

**Comment by akihikodaki** [comment]
> @akihikodaki you expect the test to fail because the signature is wrong ? let me try to repro this

No, the wrong signature is my mistake. I pushed commit 390f0d80c653fa87dac6da5c8dd1516228b0fd9b, with the correct signature. Now it succeeds if `useCompiledXaml` is true, but it still fails if `useCompiledXaml` is false.
